File: reference/android/src/test/java/github/penguin/reference/TypeChannelTest.java
package github.penguin.reference;

import org.junit.Before;
import org.junit.Test;

import java.util.Collections;

import github.penguin.reference.TestClasses.TestClass;
import github.penguin.reference.TestClasses.TestListener;
import github.penguin.reference.reference.PairedInstance;
import github.penguin.reference.reference.TypeChannel;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class TypeChannelTest {
  private static TestClasses.TestMessenger testManager;
  private static TypeChannel<TestClasses.TestClass> testChannel;

  @Before
  public void setUp() {
    testManager = new TestClasses.TestMessenger();
    testChannel = new TypeChannel<>(testManager, "test_channel");
  }

  @Test
  public void createNewInstancePair() {
    final TestClasses.TestClass testClass = new TestClasses.TestClass(testManager);
    final TestListener<PairedInstance> testListener = new TestListener<>();

    testChannel.createNewInstancePair(testClass).setOnCompleteListener(testListener);
    assertEquals(new PairedInstance("test_instance_id"), testListener.result);

    testChannel.createNewInstancePair(testClass).setOnCompleteListener(testListener);
    assertNull(testListener.result);

    assertTrue(testManager.isPaired(testClass));

    testChannel.createNewInstancePair(testClass, new Object()).setOnCompleteListener(testListener);
    assertNull(testListener.result);
  }

  @Test
  public void invokeStaticMethod() {
    final TestListener<Object> testListener = new TestListener<>();
    testChannel.invokeStaticMethod("aStaticMethod", Collections.emptyList()).setOnCompleteListener(testListener);
    assertEquals("return_value", testListener.result);
  }

  @Test
  public void invokeMethod() {
    final TestClass testClass = new TestClass(testManager);
    testChannel.createNewInstancePair(testClass);

    final TestListener<Object> testListener = new TestListener<>();

    testChannel.invokeMethod(testClass, "aMethod", Collections.emptyList()).setOnCompleteListener(testListener);
    assertEquals("return_value", testListener.result);
  }

  @Test
  public void invokeMethodOnUnpairedInstance() {
    final TestListener<Object> testListener = new TestListener<>();

    testChannel.invokeMethod(new TestClass(testManager), "aMethod", Collections.emptyList()).setOnCompleteListener(testListener);
    assertEquals("return_value", testListener.result);
  }

  @Test
  public void disposeInstancePair() {
    final TestClass testClass = new TestClass(testManager);
    final TestListener<Void> testListener = new TestListener<>();

    testChannel.createNewInstancePair(testClass);
    testChannel.disposeInstancePair(testClass).setOnCompleteListener(testListener);
    assertFalse(testManager.isPaired(testClass));

    // Test that this completes with second call.
    testChannel.disposeInstancePair(testClass).setOnCompleteListener(testListener);

    final Object owner = new Object();
    testChannel.createNewInstancePair(testClass, owner);
    testChannel.disposeInstancePair(testClass).setOnCompleteListener(testListener);
    assertTrue(testManager.isPaired(testClass));
    testChannel.disposeInstancePair(testClass, owner).setOnCompleteListener(testListener);
    assertFalse(testManager.isPaired(testClass));
  }
}

File: packages/reference/test/type_channel_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:reference/reference.dart';

import 'test_classes.dart';

void main() {
  group('$TypeChannelMessenger', () {
    late TestMessenger testMessenger;

    setUp(() {
      testMessenger = TestMessenger();
    });

    test('onReceiveCreateNewPair', () {
      expect(
        testMessenger.onReceiveCreateNewInstancePair(
          'test_channel',
          const PairedInstance('test_id'),
          <Object>[],
          owner: true,
        ),
        testMessenger.testHandler.testClassInstance,
      );
      expect(
        testMessenger.isPaired(testMessenger.testHandler.testClassInstance),
        isTrue,
      );
      expect(
        () => testMessenger.onReceiveCreateNewInstancePair(
          '',
          const PairedInstance('test_id'),
          <Object>[],
          owner: true,
        ),
        throwsAssertionError,
      );
    });

    test('onReceiveInvokeStaticMethod', () {
      expect(
        testMessenger.onReceiveInvokeStaticMethod(
          'test_channel',
          'aStaticMethod',
          <Object>[],
        ),
        'return_value',
      );
    });

    test('onReceiveInvokeMethod', () {
      testMessenger.onReceiveCreateNewInstancePair(
        'test_channel',
        const PairedInstance('test_id'),
        <Object>[],
        owner: true,
      );

      expect(
        testMessenger.onReceiveInvokeMethod(
          'test_channel',
          const PairedInstance('test_id'),
          'aMethod',
          <Object>[],
        ),
        'return_value',
      );
    });

    test('onReceiveDisposeInstancePair', () {
      testMessenger.onReceiveCreateNewInstancePair(
        'test_channel',
        const PairedInstance('test_id'),
        <Object>[],
        owner: true,
      );
      testMessenger.onReceiveDisposeInstancePair(
        const PairedInstance('test_id'),
      );
      expect(
        testMessenger.isPaired(testMessenger.testHandler.testClassInstance),
        isFalse,
      );
    });
  });

  group('$TypeChannel', () {
    late TestMessenger testMessenger;
    late TypeChannel<TestClass> testChannel;

    setUp(() {
      testMessenger = TestMessenger();
      testChannel = TypeChannel<TestClass>(testMessenger, 'test_channel');
    });

    test('createNewInstancePair', () {
      final TestClass testClass = TestClass();

      expect(
        testChannel.createNewInstancePair(
          testClass,
          owner: true,
        ),
        completion(const PairedInstance('test_reference_id')),
      );
      expect(testMessenger.isPaired(testClass), isTrue);

      expect(
        testChannel.createNewInstancePair(testClass, owner: true),
        completion(isNull),
      );
      expect(testMessenger.isPaired(testClass), isTrue);

      expect(
        testChannel.createNewInstancePair(testClass, owner: true),
        completion(isNull),
      );
      expect(testMessenger.isPaired(testClass), isTrue);
    });

    test('invokeStaticMethod', () {
      expect(
        testChannel.sendInvokeStaticMethod('aStaticMethod', <Object>[]),
        completion('return_value'),
      );
    });

    test('invokeMethod', () {
      final TestClass testClass = TestClass();

      testChannel.createNewInstancePair(testClass, owner: true);
      expect(
        testChannel.sendInvokeMethod(testClass, 'aMethod', <Object>[]),
        completion('return_value'),
      );
    });

    test('disposeInstancePair', () {
      final testClass = TestClass();

      testChannel.createNewInstancePair(testClass, owner: true);
      expect(testChannel.disposeInstancePair(testClass), completes);
      expect(testMessenger.isPaired(testClass), isFalse);

      expect(testChannel.disposeInstancePair(testClass), completes);

      // Test that this completes with second call.
      expect(testChannel.disposeInstancePair(testClass), completes);
    });
  });
}
